### PR TITLE
Use integers in Lua 5.3

### PR DESCRIPTION
**Pull request by v1993**
This patch utilizes Lua functions for integers, making them appear the right way on Lua 5.3.

I've tested this on a few of configurations (all use XUbuntu 18.04):

* 64-bit, Lua 5.3
* 64-bit, Lua 5.2
* 64-bit, Lua 5.1
* 32-bit, Lua 5.3
* 32-bit, Lua 5.1 (added)

There are few tings to remember:

* This changes behavior on Lua 5.3 for numbers which don't have integer representation: they now cause error instead of silently rounding towards zero. Sounds reasonable for me.
* GCC show warning about int-to-pointer conversion on Lua 5.3 32-bit platform. Doesn't sound significant to me anyways as everything works.

<s>I probably should test on Lua 5.1 32-bit as well, will report results soon.</s> Done, it works.

### CODE REVIEW

**Comment by v1993** [comment]
Works on 32-bit Lua 5.1 too.

**Comment by v1993** [comment]
@psychon @pavouk What do you think about this PR?

File: lgi/lgi.h [hunk 1]
```diff
@@ -29,6 +29,29 @@
 #endif
 #endif
 
+/* Lua 5.3 integers. */
+#if LUA_VERSION_NUM >= 503
+// For Lua 5.3
+typedef LUA_UNSIGNED lgi_Unsigned;
```
Comment (psychon) [inline lgi/lgi.h]: Why `LUA_UNSIGNED`? It does not appear in the API documentation of Lua 5.3. Only `lua_Unsigned` does. Also, the Lua 5.1 and Lua 5.2 cases use `lua_Unsigned`.

File: lgi/lgi.h [hunk 2]
```diff
@@ -29,6 +29,29 @@
...
+#if LUA_INT_TYPE == LUA_INT_INT
+#define LGI_LUAINT_FORMAT "%d"
+#elseif LUA_INT_TYPE == LUA_INT_LONG
+#define LGI_LUAINT_FORMAT "%ld"
+#else
+#define LGI_LUAINT_FORMAT "%lld"
+#endif
```
Comment (psychon) [inline lgi/lgi.h]: All of these `#ifdef`s can be replaced by just using `LUA_INTEGER_FMT`, I think (but haven't checked).

File: lgi/lgi.h [hunk 3]
```diff
@@ -29,6 +29,29 @@
...
+#elif LUA_VERSION_NUM == 502
+// For Lua 5.2
+typedef lua_Unsigned lgi_Unsigned;
+#define lua_isinteger(L, i) 0
+#define LGI_LUAINT_FORMAT "%td"
```
Comment (psychon) [inline lgi/lgi.h]: I had to check the man page of `printf`: Why do you think that this is `ptrdiff_t` argument?

File: lgi/marshal.c [hunk 4]
```diff
@@ -68,16 +83,21 @@ marshal_2c_int (lua_State *L, GITypeTag tag, GIArgument *val, int narg,
...
-      HANDLE_INT(INT32, int32, INT, gint, -0x80000000LL, 0x7fffffffLL, s);
-      HANDLE_INT(UINT32, uint32, UINT, guint, 0, 0xffffffffUL, u);
-      HANDLE_INT(UNICHAR, uint32, UINT, guint, 0, 0x7fffffffLL, u);
+      HANDLE_INT(INT8, int8, INT, gint, G_MININT8, G_MAXINT8, s);
+      HANDLE_INT(UINT8, uint8, UINT, guint, 0, G_MAXUINT8, u);
+      HANDLE_INT(INT16, int16, INT, gint, G_MININT16, G_MAXINT16, s);
+      HANDLE_INT(UINT16, uint16, UINT, guint, 0, G_MAXUINT16, u);
+      HANDLE_INT(INT32, int32, INT, gint, G_MININT32, G_MAXINT32, s);
+      HANDLE_INT(UINT32, uint32, UINT, guint, 0, G_MAXUINT32, u);
+      HANDLE_INT(UNICHAR, uint32, UINT, guint, 0, G_MAXUINT32, u);
```
Comment (psychon) [inline lgi/marshal.c]: Why are you replacing the constants with glib macros here (and not in the `#else` below)? That seems like an unrelated change. Or was there something wrong with them?

File: lgi/marshal.c [hunk 5]
```diff
@@ -112,7 +132,7 @@ marshal_2lua_int (lua_State *L, GITypeTag tag, GIArgument *val,
 	    ReturnUnion *ru = (ReturnUnion *) val;			\
 	    ru->arg.v_ ## namelower = (g ## namelower) ru->ut;		\
 	  }								\
-	lua_pushnumber (L, parent == LGI_PARENT_FORCE_POINTER		\
+	lua_pushinteger (L, parent == LGI_PARENT_FORCE_POINTER		\
```
Comment (psychon) [inline lgi/marshal.c]: This change assumes that the Lua integer type has at least 64 bits. Actually... since `double` only has something like a 53 bits mantissa, I guess that (u)int64 passing never really worked correctly anyway, so this can be ignored. Right?

File: tests/gireg.lua [hunk 6]
```diff
@@ -16,6 +19,9 @@ local check = testsuite.check
 local checkv = testsuite.checkv
 local gireg = testsuite.group.new('gireg')
 
+-- Do we have Lua-side support for integers?
+local nativeIntegers = (_VERSION == 'Lua 5.3' or _VERSION == 'Lua 5.4')
```
Comment (psychon) [inline tests/gireg.lua]: `_VERSION >= 'Lua 5.3`

File: tests/gireg.lua [hunk 7]
```diff
@@ -36,8 +42,10 @@ function gireg.type_int8()
    checkv(R.test_int8(0), 0, 'number')
    checkv(R.test_int8(1), 1, 'number')
    checkv(R.test_int8(-1), -1, 'number')
-   checkv(R.test_int8(1.1), 1, 'number')
-   checkv(R.test_int8(-1.1), -1, 'number')
+   if not nativeIntegers then
```
Comment (psychon) [inline tests/gireg.lua]: Would be nice to leave a comment here explaining why these are skipped. Or alternatively, since I expect that these will error out, it could be possible to add an `else` that checks via `pcall` for the error.

However, this raises the question on consistency: Some LGI versions now silently round while others error out. Should `check_integer` for Lua < 5.3 also check for integer values and throw an error? That way, behaviour would always be (more or less) the same.

Alternatively, the safer version would be to keep this rounding behaviour. That way the likelyhood of breaking someone's code is smaller.

File: tests/gireg.lua [hunk 8]
```diff
@@ -154,18 +174,28 @@ function gireg.type_int64()
...
---   checkv(R.test_int64(0x7fffffffffffffff), 0x7fffffffffffffff, 'number')
---   checkv(R.test_int64(-0x8000000000000000), -0x8000000000000000, 'number')
---   check(not pcall(R.test_int64, 0x8000000000000000))
---   check(not pcall(R.test_int64, -0x8000000000000001))
+-- UPDATE: It IS working in Lua 5.3 because of its integer support,
+-- so tests are enabled there.
+
+   if nativeIntegers then
+      checkv(R.test_int64(0x7fffffffffffffff), 0x7fffffffffffffff, 'number')
+      checkv(R.test_int64(-0x8000000000000000), -0x8000000000000000, 'number')
```
Comment (psychon) [inline tests/gireg.lua]: Can this even work correctly? The constant before `-` is applied is larger than the maximum value a signed 64 bit integer can hold, so this should overflow. I doubt that Lua guarantees anything in the presence of integer overflows...

File: tests/gireg.lua [hunk 9]
```diff
@@ -154,18 +174,28 @@ function gireg.type_int64()
...
+   end
+
+   -- Following two lines succeed because integers wrap around in Lua 5.3 like they do in C.
```
Comment (psychon) [inline tests/gireg.lua]: Integers *do not* wrap around in C. Please remove that remark.
(I wonder how Lua manages to get wrap-around integers... From a quick look at the implementation: It doesn't.)

File: lgi/lgi.h (hunk 1 above)
Comment (v1993) [reply lgi/lgi.h]: I saw this in `luaconf.h`, so came with this. Ok, going to fix.

File: lgi/lgi.h (hunk 2 above)
Comment (v1993) [reply lgi/lgi.h]: Going to check on Lua 5.1/5.2.

File: lgi/marshal.c (hunk 4 above)
Comment (v1993) [reply lgi/marshal.c]: I was changing this part of code anyways, so was thinking that hardcoding stuff is generally bad. Result works flawlessly.

As for the `#else` clause, I left those because I'm not sure about pre-integers 64-bits stuff.

File: tests/gireg.lua (hunk 6 above)
Comment (v1993) [reply tests/gireg.lua]: This seems a bit complex as `_VERSION` is a string, but probably then

```Lua
local nativeIntegers = math.maxinteger
```

as we really care about integers support and not Lua version?

File: tests/gireg.lua (hunk 7 above)
Comment (v1993) [reply tests/gireg.lua]: That's something I'm not 100% sure about. I think that throwing error on non-integer is good idea (that's what Lua itself does).

Whole change is done inside Lua core itself, so I'm don't really think that it's problem of lgi. Also, there are few places where < Lua 5.3 code actually use `lua_tonumber` on integer values.

File: tests/gireg.lua (hunk 9 above)
Comment (v1993) [reply tests/gireg.lua]: What am I missing?

```Lua
Lua 5.3.3  Copyright (C) 1994-2016 Lua.org, PUC-Rio
> 0x7fffffffffffffff
9223372036854775807
> 0x7fffffffffffffff+1
-9223372036854775808
```

They do wrap just fine…

**Comment by v1993** [comment]
I'm currently a little busy, will get back on this in day or two.

**Comment by v1993** [comment]
@pavouk @psychon Can somebody take a look at this, please?

**Comment by psychon** [comment]
Sorry for the long delay.

> Does lgi support windows? If so, you may wish to test there as well both with 32 and 64 bit builds.

In theory: yes (I guess). IIRC, the Meson build was tested on Windows.
However, I have no way to test the windows build.
